README.md
# Squiss [![Build Status](https://travis-ci.org/TomFrost/Squiss.svg?branch=master)](https://travis-ci.org/TomFrost/Squiss)
High-volume Amazon SQS Poller and single-queue client for Node.js 4 and up

```javascript
const poller = new Squiss({
  queueName: 'my-sqs-queue',
  bodyFormat: 'json',
  unwrapSns: true,
  maxInFlight: 500
});
poller.start();

poller.on('message', (msg) => {
  console.log('%s says: %s', msg.body.name, msg.body.message);
  msg.del();
});
```

## How it works
Squiss aims to process as many messages simultaneously as possible. Set the `maxInFlight` option to the number of messages your app can handle at one time without choking, and Squiss will attempt to keep that many messages flowing through your app, grabbing more as you mark each message as handled or ready for deletion. If the queue is empty, Squiss will maintain an open connection to SQS, waiting for any messages that appear in real time. By default, anyway. You can configure it to poll however you want. Squiss don't care.

## Functions

### new Squiss(opts)
Don't be scared of `new` -- you need to create a new Squiss instance for every queue you want to poll. Squiss is an EventEmitter, so don't forget to call `squiss.on('message', (msg) => msg.del())` at the very least.

#### opts {...}
Use the following options to point Squiss at the right queue:
- **opts.awsConfig** An object mapping to pass to the SQS constructor, configuring the aws-sdk library. This is commonly used to set the AWS region, endpoint, or the user credentials. See the docs on [configuring the aws-sdk](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html) for details.
- **opts.queueName** The name of the queue to be polled. Used only if opts.queueUrl is not specified, but Squiss prefers just the name.
- **opts.queueUrl** The URL of the queue to be polled. If not specified, opts.queueName is required.
- **opts.accountNumber** If a queueName is specified, the accountNumber of the queue owner can optionally be specified to access a queue in a different AWS account.
- **opts.correctQueueUrl** _Default false._ Changes the protocol, host, and port of the queue URL to match the configured SQS endpoint (see opts.awsConfig), applicable only if opts.queueName is specified. This can be useful for testing against a local SQS service, such as ElasticMQ.

Squiss's defaults are great out of the box for most use cases, but you can use the below to fine-tune your Squiss experience:
- **opts.SQS** _Default AWS.SQS_ An instance of the official SQS Client, or an SQS constructor function to use rather than the default one provided by AWS.SQS
- **opts.activePollIntervalMs** _Default 0._ The number of milliseconds to wait between requesting batches of messages when the queue is not empty, and the maxInFlight cap has not been hit. For most use cases, it's better to leave this at 0 and let Squiss manage the active polling frequency according to maxInFlight.
- **opts.advancedCallMs** _Default 5000._ If `opts.autoExtendTimeout` is used, this is the number of milliseconds that Squiss will make the call to extend the VisibilityTimeout of the message before the message is set to expire.
- **opts.autoExtendTimeout** _Default false._ If true, Squiss will automatically extend each message's VisibilityTimeout in the SQS queue until it's handled (by keeping, deleting, or releasing it). It will place the API call to extend the timeout `opts.advancedCallMs` milliseconds in advance of the expiration, and will extend it by the number of seconds specified in `opts.visibilityTimeoutSecs`. If that's not specified, the VisibilityTimeout setting on the queue itself will be used.
- **opts.bodyFormat** _Default "plain"._ The format of the incoming message. Set to "json" to automatically call `JSON.parse()` on each incoming message.
- **opts.deleteBatchSize** _Default 10._ The number of messages to delete at one time. Squiss will trigger a batch delete when this limit is reached, or when deleteWaitMs milliseconds have passed since the first queued delete in the batch; whichever comes first. Set to 1 to make all deletes immediate. Maximum 10.
- **opts.deleteWaitMs** _Default 2000._ The number of milliseconds to wait after the first queued message deletion before deleting the message(s) from SQS.
- **opts.idlePollIntervalMs** _Default 0._ The number of milliseconds to wait before requesting a batch of messages when the queue was empty on the prior request.
- **opts.maxInFlight** _Default 100._ The number of messages to keep "in-flight", or processing simultaneously. When this cap is reached, no more messages will be polled until currently in-flight messages are marked as deleted or handled. Setting this option to 0 will uncap your inFlight messages, pulling and delivering messages as long as there are messages to pull.
- **opts.noExtensionsAfterSecs** _Default 43200._ If `opts.autoExtendTimeout` is used, Squiss will stop auto-renewing a message's VisibilityTimeout when it reaches this age. Default is 12 hours, SQS's VisbilityTimeout maximum.
- **opts.pollRetryMs** _Default 2000._ The number of milliseconds to wait before retrying when Squiss's call to retrieve messages from SQS fails.
- **opts.receiveBatchSize** _Default 10._ The number of messages to receive at one time. Maximum 10 or maxInFlight, whichever is lower.
- **opts.receiveWaitTimeSecs** _Default 20._ The number of seconds for which to hold open the SQS call to receive messages, when no message is currently available. It is recommended to set this high, as Squiss will re-open the receiveMessage HTTP request as soon as the last one ends. If this needs to be set low, consider setting activePollIntervalMs to space out calls to SQS. Maximum 20.
- **opts.unwrapSns** _Default false._ Set to `true` to denote that Squiss should treat each message as though it comes from a queue subscribed to an SNS endpoint, and automatically extract the message from the SNS metadata wrapper.
- **opts.visibilityTimeoutSecs** _Defaults to queue setting on read, or 30 seconds for createQueue._ The amount of time, in seconds, that received messages will be unavailable to other pollers without being deleted.

Are you using Squiss to create your queue, as well? Squiss will use `opts.receiveWaitTimeSecs` and `opts.visibilityTimeoutSecs` above in the queue settings, but consider setting any of the following options to configure it further. Note that the defaults are the same as Amazon's own:
- **opts.delaySecs** _Default 0._ The number of milliseconds by which to delay the delivery of new messages into the queue by default.
- **opts.maxMessageBytes** _Default 262144 (256KB)._ The maximum size of a single message, in bytes, that the queue can support.
- **opts.messageRetentionSecs** _Default 345600 (4 days)._ The amount of time for which to retain messages in the queue until they expire, in seconds. Maximum is 1209600 (14 days).
- **opts.queuePolicy** If specified, will be set as the access policy of the queue when `createQueue` is called. See [the AWS Policy documentation](http://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) for more information.

### squiss.createQueue()
Creates the configured queue! This returns a promise that resolves with the new queue's URL when it's complete. Note that this can only be called if you set `opts.queueName` when instantiating Squiss.

### squiss.deleteMessage(Message)
Deletes a message, given the full Message object sent to the `message` event. It's much easier to call `message.del()`, but if you need to do it right from the Squiss instance, this is how. Note that the message probably won't be deleted immediately -- it'll be queued for a batch delete. See the constructor notes for how to configure the specifics of that.

### squiss.changeMessageVisibility(Message|receiptHandle, timeoutInSeconds)
Changes the visibility timeout of a message, given either the full Squiss Message object or the receipt handle string.

### squiss.deleteQueue()
Deletes the configured queue, returning a promise that resolves on complete. Squiss lets you do this, even though it makes Squiss useless. Squiss is so selfless.

### squiss.getQueueUrl()
Returns a Promise that resolves with the URL of the configured queue, even if you only instantiated Squiss with a queueName. The correctQueueUrl setting applies to this result, if it was set.

### squiss.handledMessage(Message)
Informs Squiss that you got a message that you're not planning on deleting, so that Squiss can decrement the number of "in-flight" messages. It's good practice to delete every message you process, but this can be useful in case of error. You can also call `message.keep()` on the message itself to invoke this.

### squiss.releaseMessage(Message)
Releases the given Message object back to the queue by setting its `VisibilityTimeout` to `0` and marking the message as handled internally. You can also call `message.release()` on the message itself to invoke this.

### squiss.sendMessage(message, delay, attributes)
Sends an individual message to the configured queue, and returns a promise that resolves with AWS's official message metadata: an object containing `MessageId`, `MD5OfMessageAttributes`, and `MD5OfMessageBody`. Arguments:
- **message**. The message to push to the queue. If it's a string, great! If it's an Object, Squiss will call JSON.stringify on it.
- **delay** _optional_. The amount of time, in seconds, to wait before making the message available in the queue. If not specified, the queue's configured value will be used.
- **attributes** _optional_. An optional attributes mapping to associate with the message. For more information, see [the official AWS documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SQS.html#sendMessage-property).

### squiss.sendMessages(messages, delay, attributes)
Sends an array of any number of messages to the configured SQS queue, breaking them down into appropriate batch requests executed in parallel (or as much as the default HTTP agent allows). It returns a promise that resolves with a response closely aligned to the official AWS SDK's sendMessageBatch, except the results from all batch requests are merged. Expect a result similar to:

```javascript
{
  Successful: [
    {Id: string, MessageId: string, MD5OfMessageAttributes: string, MD5OfMessageBody: string}
  ],
  Failed: [
    {Id: string, SenderFault: boolean, Code: string, Message: string}
  ]
}
```

The "Id" supplied in the response will be the index of the message in the original messages array, in string form. Arguments:
- **messages**. The array of messages to push to the queue. The messages should be either strings, or Objects that Squiss can pass to JSON.stringify.
- **delay** _optional_. The amount of time, in seconds, to wait before making the messages available in the queue. If not specified, the queue's configured value will be used.
- **attributes** _optional_. An optional attributes mapping to associate with each message. For more information, see [the official AWS documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SQS.html#sendMessage-property).

### squiss.start()
Starts polling SQS for new messages. Each new message is handed off in the `message` event.

### squiss.stop(soft=`false`)
Hold on to your hats, this one stops the polling, aborting any in-progress request for new messages. If called with soft=`true` while there's an active request for new messages, the active request will not be aborted and the message event may still be fired up to `opts.receiveWaitTimeSecs` afterward.

## Properties

### {number} squiss.inFlight
The number of messages currently in-flight.

### {boolean} squiss.running
`true` if Squiss is actively polling SQS. If it's not polling, we made the genius design decision to have this set to `false`.

### {Object} squiss.sqs
For your convenience, Squiss provides direct access to the AWS SDK's SQS object, which can be handy for setting up or tearing down tests. No need to thank Squiss. Squiss does this because Squiss cares.

## Events

### deleted {Message}
Emitted when a message is confirmed as being successfully deleted from the queue. The `handled` and `delQueued` events will also be fired for deleted messages, but that will come earlier, when the delete function is initially called.

### delError {Object}
A `delError` is emitted when AWS reports that any of the deleted messages failed to actually delete. The
object handed to you in this event is the AWS failure object described in the [SQS deleteMessageBatch documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SQS.html#getQueueUrl-property).

### delQueued {Message}
Emitted when a message is queued for deletion, even if delete queuing has been turned off.

### drained
Emitted when the last in-flight message has been handled, and there are no more messages currently in flight.

### error {Error}
If any of the AWS API calls outrightly fail, `error` is emitted. If you don't have a listener on `error`, per Node.js's structure, the error will be treated as uncaught and will crash your app.

### gotMessages {number}
Emitted when Squiss asks SQS for a new batch of messages, and gets some (or one). Supplies the number of retrieved messages.

### handled {Message}
Emitted when a message is handled by any means: deleting, releasing, or calling `keep()` or `handledMessage()` on it. 

### queueEmpty
Emitted when Squiss asks SQS for new messages, and doesn't get any.

### released {Message}
Emitted after `release()` or `releaseMessage` has been called and the VisibilityTimeout of a message has successfully been changed to `0`. The `handled` event will also be fired for released messages, but that will come earlier, when the release function is initially called.
 
### timeoutExtended {Message}
Emitted when a message has had its timeout successfully extended by the `autoExtendTimeout` feature.

### aborted
Emitted after a hard stop() if a request for new messages was already in progress.

### autoExtendFail {message: Message, error: Error}
Emitted if automatic VisibilityTimeout extensions are enabled, and Squiss attempts to extend a message that has either been deleted or otherwise expired.

### maxInFlight
Emitted when Squiss has hit the maxInFlight cap. At this point, Squiss won't retrieve any more messages until at least `opts.receiveBatchSize` in-flight messages have been deleted.

### message {Message}
Emitted every time Squiss pulls a new message from the queue. The Squiss Message object handed back has the following methods and properties:

#### {Object|string} message.body
The body of the SQS message, unwrapped from the SNS metadata wrapper (if `unwrapSns` was specified in the constructor), and JSON-parsed (if `bodyFormat: 'json'` was specified in the constructor). Otherwise the body will just be a string.

#### {string} message.subject
The subject of the SNS message, if set. Exists only if unwrapSns was specified.

#### {string} message.topicArn
The full SNS topic ARN to which this message was posted. Exists only if unwrapSns was specified.

#### {string} message.topicName
The name of the SNS topic to which this message was posted. Included for convenience only: this is the last segment of the topicArn. Exists only if unwrapSns was specified.

#### message.del()
Deletes the message from SQS. Either this or `message.keep()` _must_ be called on each message Squiss delivers in order to maintain an accurate inFlight count.

#### message.keep()
Instructs Squiss that you're not planning to delete a message, but it should no longer be considered "in-flight". Either this, `message.del()`, or `message.release()` _must_ be called on each message Squiss delivers in order to maintain an accurate inFlight count.

#### message.release()
Changes the visibility timeout of the message to 0.

#### message.changeVisibility(timeoutInSeconds)
Changes the visibility timeout of the message.

#### {Object} message.raw
The raw, unprocessed SQS response object as delivered from the aws-sdk.

## Versions
Squiss supports Node 4 LTS and higher. For 0.12 support, consider compiling with Babel or using Squiss version 0.x.

## License
Squiss is Copyright (c) 2017 Tom Shawver, released under the ultra-permissive ISC license. See LICENSE.txt for details.

## Credits
Squiss was originally created at [TechnologyAdvice](http://www.technologyadvice.com) in Nashville, TN.



CHANGELOG.md
# Squiss Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## [Development]
Nothing yet!

## [v2.2.1]
### Fixed
- The `advancedCallMs` option had no effect unless passed to TimeoutExtender directly, which no sane person would do.
- The `noExtensionsAfterSecs` option defaulted to zero, which led to messages never being auto-extended unless a custom value was passed. 

## [v2.2.0]
### Added
- The `advancedCallMs` option to tweak how far in advance of a message's VisibilityTimeout that a ChangeVisibilityTimeout call will be made

### Fixed
- A possible scenario where someone could set the visibilityTimeoutSecs ultra low and cause timers to be set for negative milliseconds

## [v2.1.1]
### Fixed
- The option `noExtensionsAfterSecs` now defaults to seconds instead of milliseconds

## [v2.1.0]
### Added
- The `autoExtendFail` event, fired when the TimeoutExtender fails to change the VisibilityTimeout of a message due to the message no longer existing or being expired. This event may encompass more errors in the future.

### Fixed
- TimeoutExtender no longer continually tries to extend messages that SQS has reported as previously deleted or expired. Instead, messages are removed from the list on this error, and a `autoExtendFail` event is emitted.
- TimeoutExtender no longer tries to extend the timeout beyond AWS's maximum message lifetime.

## [v2.0.2]
### Fixed
- Message now passes an instance of itself to the handled call, fixing cases where the message would not be emitted with the "handled" event.

## [v2.0.1]
### Fixed
- TimeoutExtender now passes seconds to `changeMessageVisibility` instead of milliseconds. AWS didn't like that.

## [v2.0.0]
### Added
- The `deleted` event, fired when a message has been confirmed to be successfully deleted
- The `delQueued` event, fired when a message deletion has been queued up for the next batch
- The `handled` event, fired when a message has been handled by any means (kept, deleted, or released)
- The `released` event, fired when a message has been confirmed to be successfully released back for immediate availability
- The TimeoutExtender feature, which can automatically extend the VisibilityTimeout of any message that has not yet been handled. See README for details!
- The `autoExtendTimeout` and `noExtensionsAfterSecs` options to support the TimeoutExtender feature
- The `getQueueVisibilityTimeout` function. Intuitively, this gets the queue's VisibilityTimeout.
- `opts.SQS` now supports an instance of the SQS client class in addition to a constructor (cwhenderson20)

### Changed
- BREAKING: The `deleteMessage` method now requires a Message object to be passed to it; not just the receiptHandle
- BREAKING: The `releaseMessage` method now requires a Message object to be passed to it; not just the receiptHandle
- BREAKING: The `handledMessage` method now requires a Message object to be passed to it
- The `start()` function now returns a Promise that resolves when the poller has started

## [v1.1.0]
### Added
- You can now pass an SQS constructor from your own aws-sdk version instead of using the one packaged with Squiss (cwhenderson20)
- The visibilityTimeout of a received message can now be changed with `squiss.changeMessageVisibility` and `message.changeVisibility` (tothandras)
- A message can now be returned to the queue and made immediately available with `message.release` (tothandras)
- Messages can also be returned to the queue with immediate availablility through `squiss.releaseMessage`

### Fixed
- Deleting messages by handle now works. The Message ID used will be the position of the message in the delete queue, 1-indexed. ([#13](https://github.com/TomFrost/Squiss/issues/13))

### Changed
- Tests now use dirty-chai to work around Chai's cardinal sin of allowing nonexistent assertions to be called without error
- Tests now use chai-as-promised to assert promise results

## [v1.0.0]
### Added
- The `gotMessages` event, which fires when Squiss retrieves a new batch of messages from SQS
- The `maxInFlight` event, which fires when Squiss stops requesting new messages due to hitting the maxInFlight cap
- The `aborted` event, which fires when stop() is called during an active SQS receive message request
- Documentation for the already-existing `drained` event
- Documentation for the already-existing `sqs` property
- New option: `activePollIntervalMs`, allowing SQS requests to be spaced out when the queue has messages
- New option: `idlePollIntervalMs`, allowing SQS requests to be spaced out when the queue is empty
- `getQueueUrl` method to retrieve the configured queue's URL, even if only the name was provided to the constructor
- `createQueue` method to create the configured queue
- `deleteQueue` method to delete the configured queue
- `sendMessage` method to send a message to the configured queue
- `sendMessages` method to send an array of messages of any size (within reason) to the configured queue
- `deleteMessage` now accepts a ReceiptHandle string in lieu of a Message object, allowing messages to be deleted later without caching the full message itself.

### Changed
- Dropped support for Node 0.12. For Node 0.12 support, consider compiling with an ES6 transpiler such as Babel, or using version 0.7.
- Switched codebase to native ES6, updating to newest TechnologyAdvice style guide
- `stop()` now aborts any ongoing receiveMessage request by default. It also accepts a `soft` argument (boolean) to soft-stop the poller without the abort functionality.
- `opts.visibilityTimeout` has been renamed to `opts.visibilityTimeoutSecs` for consistency.

## [v0.7.0]
### Added
- pollRetryMs has been added to automatically restart the poller after a failed call to get new messages (dannyrscott)

## [v0.6.0]
### Added
- maxInFlight can now be set to 0 to uncap message flow (dannyrscott)

### Fixed
- In low-volume use cases, the delete timer would only fire once, and all other deletes depended on filling the queue (dannyrscott)


test/setup.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

'use strict'

const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')
const dirtyChai = require('dirty-chai')
const path = require('path')
const mod = require('module')
const sinon = require('sinon')
const sinonChai = require('sinon-chai')

chai.use(sinonChai)
chai.use(chaiAsPromised)
chai.use(dirtyChai)

global.should = chai.should()
global.sinon = sinon

// importing files with ../../../../../.. makes my brain hurt
process.env.NODE_PATH = path.join(__dirname, '..') + path.delimiter + (process.env.NODE_PATH || '')
mod._initPaths()


test/index.spec.js
import chai from 'chai';
const expect = chai.expect;

describe('index', () => {});

test/Message.spec.js
import chai from 'chai';
const expect = chai.expect;

describe('Message', () => {});

src/Message.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

'use strict'

/**
 * The message class is a wrapper for Amazon SQS messages that provides the raw and parsed message body,
 * optionally removed SNS wrappers, and provides convenience functions to delete or keep a given message.
 */
class Message {

  /**
   * Creates a new Message.
   * @param {Object} opts A mapping of message-creation options
   * @param {Object} opts.msg A parsed SQS response as returned from the official aws-sdk
   * @param {boolean} [opts.unwrapSns=false] Set to `true` to denote that each message should be treated as though
   *    it comes from a queue subscribed to an SNS endpoint, and automatically extract the message from the SNS
   *    metadata wrapper.
   * @param {string} opts.bodyFormat "plain" to not parse the message body, or "json" to pass it through JSON.parse
   *    on creation
   * @param {Squiss} opts.squiss The squiss instance responsible for retrieving this message. This will be used to
   *    delete the message and update inFlight count tracking.
   */
  constructor(opts) {
    this.raw = opts.msg
    this.body = opts.msg.Body
    if (opts.unwrapSns) {
      let unwrapped = JSON.parse(this.body)
      this.body = unwrapped.Message
      this.subject = unwrapped.Subject
      this.topicArn = unwrapped.TopicArn
      this.topicName = unwrapped.TopicArn.substr(unwrapped.TopicArn.lastIndexOf(':') + 1)
    }
    this.body = Message._formatMessage(this.body, opts.bodyFormat)
    this._squiss = opts.squiss
    this._handled = false
  }

  /**
   * Queues this message for deletion.
   */
  del() {
    if (!this._handled) {
      this._squiss.deleteMessage(this)
      this._handled = true
    }
  }

  /**
   * Keeps this message, but releases its inFlight slot in Squiss.
   */
  keep() {
    if (!this._handled) {
      this._squiss.handledMessage(this)
      this._handled = true
    }
  }

  /**
   * Changes the visibility timeout of the message to 0.
   */
  release() {
    if (!this._handled) {
      this._squiss.releaseMessage(this)
      this._handled = true
    }
  }

  /**
   * Changes the visibility timeout of the message.
   */
  changeVisibility(timeoutInSeconds) {
    return this._squiss.changeMessageVisibility(this, timeoutInSeconds)
  }
}

/**
 * Parses a message according to the given format.
 * @param {string} msg The message to be parsed
 * @param {string} format The format of the message. Currently supports "json".
 * @returns {Object|string} The parsed message, or the original message string if the format type is unknown.
 * @private
 */
Message._formatMessage = (msg, format) => {
  switch (format) {
  case 'json': return JSON.parse(msg)
  default: return msg
  }
}

module.exports = Message


src/TimeoutExtender.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

'use strict'

/**
 * The maximum age, in milliseconds, that a message can reach before AWS will no longer accept VisibilityTimeout
 * extensions.
 * @type {number}
 */
const MAX_MESSAGE_AGE_MS = 43200000

/**
 * Option defaults.
 * @type {Object}
 */
const optDefaults = {
  visibilityTimeoutSecs: 30,
  noExtensionsAfterSecs: MAX_MESSAGE_AGE_MS / 1000,
  advancedCallMs: 5000
}

/**
 * The TimeoutExtender is a module that attaches itself to a Squiss instance via event
 * listeners, and uses the instance's public API to automatically extend the
 * VisibilityTimeout of the message until it is either handled (which includes actions
 * such as deleting or releasing the message) or it reaches an age that is greater than
 * the `noExtensionsAfterSecs` option. Functionally, this class could be distributed as a
 * third party module and attached to a Squiss instance by the user, however the common
 * and convenient use case is for this class to be used internally by Squiss itself.
 *
 * For efficient operation, the TimeoutExtender combines a doubly-linked list with a
 * hash map, keeping an index of every message while not incurring the RAM or CPU overhead
 * of having to re-create an array every time a message is deleted from the middle of a
 * sorted queue. Every operation in this class has a time complexity of O(1) and a space
 * complexity of O(n).
 */
class TimeoutExtender {

  /**
   * Creates a new TimeoutExtender.
   * @param {Squiss} squiss The Squiss instance on which this extender should operate
   * @param {Object} opts An object containing options mappings
   * @param {number} [opts.visibilityTimeoutSecs=30] The queue's VisibilityTimeout, in
   * seconds. This will be used to know when to extend a message. It will also be the
   * amount of time by which the timeout gets extended.
   * @param {number} [opts.noExtensionsAfterSecs=43200] The number of seconds after which
   * a message's VisibilityTimeout should not be extended. The default and maximum value
   * (imposed by AWS) is 43200 (12 hours).
   * @param {number} [opts.advancedCallMs=5000] The amount of time before a message's
   * VisibilityTimeout expiration to make the API call to extend it. Setting this too
   * high will cause a message to be extended within a time frame where it may be deleted
   * anyway; setting too low may cause the message to expire before the API call can
   * complete. The max is `opts.visibilityTimeoutSecs * 1000`.
   */
  constructor(squiss, opts) {
    this._opts = Object.assign({}, optDefaults, opts)
    this._head = null
    this._tail = null
    this._index = {}
    this._timer = null
    this._squiss = squiss
    this._squiss.on('handled', msg => this.deleteMessage(msg))
    this._squiss.on('message', msg => this.addMessage(msg))
    this._visTimeout = this._opts.visibilityTimeoutSecs * 1000
    this._stopAfter = Math.min(this._opts.noExtensionsAfterSecs * 1000, MAX_MESSAGE_AGE_MS)
    this._apiLeadMs = Math.min(this._opts.advancedCallMs, this._visTimeout)
  }

  /**
   * Adds a new message to the tracker.
   * @param {Message} message A Squiss Message object
   */
  addMessage(message) {
    const now = Date.now()
    this._addNode({
      message,
      receivedOn: now,
      timerOn: now + this._visTimeout - this._apiLeadMs
    })
  }

  /**
   * Deletes a message from the tracker, if the message is currently being tracked.
   * @param {Message} message A Squiss Message object
   */
  deleteMessage(message) {
    const node = this._index[message.raw.MessageId]
    if (node) this._deleteNode(node)
  }

  /**
   * Adds a message wrapper node to the linked list and hash map index.
   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
   * object to be added
   * @private
   */
  _addNode(node) {
    this._index[node.message.raw.MessageId] = node
    if (!this._head) {
      this._head = node
      this._tail = node
      this._headChanged()
    } else {
      this._tail.next = node
      node.prev = this._tail
      this._tail = node
    }
  }

  /**
   * Deletes a message wrapper node from the linked list and hash map index.
   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
   * object to be removed
   * @private
   */
  _deleteNode(node) {
    const msgId = node.message.raw.MessageId
    delete this._index[msgId]
    if (this._head === node) {
      this._head = node.next
      if (this._head) delete this._head.prev
      this._headChanged()
    } else if (this._tail === node) {
      this._tail = this._tail.prev
      delete this._tail.next
    } else {
      node.prev.next = node.next
      node.next.prev = node.prev
    }
  }

  /**
   * Gets the current millisecond age of a tracked message, projected `opts.advancedCallMs`
   * into the future.
   * @param node The node to be checked
   * @returns {number} The age of the message in milliseconds
   * @private
   */
  _getNodeAge(node) {
    return Date.now() - node.receivedOn + this._apiLeadMs
  }

  /**
   * Called internally when the head of the linked list has changed in any way. This
   * function is responsible for maintaining the timer that determines the tracker's next
   * action.
   * @returns {boolean} true if a timer was set in response to the changed head; false
   * otherwise.
   * @private
   */
  _headChanged() {
    if (this._timer) clearTimeout(this._timer)
    if (!this._head) return false
    const node = this._head
    this._timer = setTimeout(() => {
      if (this._getNodeAge(node) >= this._stopAfter) return this._deleteNode(node)
      return this._renewNode(node)
    }, node.timerOn - Date.now())
    return true
  }

  /**
   * Extends the VisibilityTimeout of the message contained in the provided wrapper node,
   * and moves the node to the tail of the linked list.
   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
   * object to be renewed
   * @private
   */
  _renewNode(node) {
    const extendByMs = Math.min(this._visTimeout, MAX_MESSAGE_AGE_MS - this._getNodeAge(node))
    const extendBySecs = Math.floor(extendByMs / 1000)
    this._squiss.changeMessageVisibility(node.message, extendBySecs)
      .then(() => this._squiss.emit('timeoutExtended', node.message))
      .catch(err => {
        if (err.message.match(/Message does not exist or is not available/)) {
          this._deleteNode(node)
          this._squiss.emit('autoExtendFail', { message: node.message, error: err })
        } else {
          this._squiss.emit('error', err)
        }
      })
    this._deleteNode(node)
    node.timerOn = Date.now() + extendBySecs * 1000
    this._addNode(node)
  }
}

module.exports = TimeoutExtender

